include/aws/crt/Variant.h

```diff
@@ -0,0 +1,348 @@
+                if (this == &other)
+                {
+                    return *this;
+                }
+                if (this->m_value)
+                {
+                    DestroyUtil<Ts...>(this->m_index);
+                }
+
+                CopyAssignUtil(other);
+
+                return *this;
+            }
+
+            template<typename VariantT = ThisVariantT, EnableIfOtherIsSameVariantType<VariantT> = 1>
+            Variant &operator=(VariantT &&other)
+            {
+                if (this == &other)
+                {
+                    return *this;
+                }
+                if (this->m_value)
+                {
+                    DestroyUtil<Ts...>(this->m_index);
+                }
+
+                MoveAssignUtil(other);
+
+                return *this;
+            }
+
+            template<typename T, EnableIfOtherIsThisVariantAlternative<T> = 1>
+            bool holds_alternative() const
+            {
+                assert(m_value);
+                assert(m_index != -1);
+                return m_index == VariantDetail::Index::GetIndexOf<T, Ts...>();
+            }
+
+            template<typename T, EnableIfOtherIsThisVariantAlternative<T> = 1>
+            const T& get() const
+            {
+                assert(holds_alternative<T>);
+                return *reinterpret_cast<T *>(m_value);
+            }
+
+            template<typename T, EnableIfOtherIsThisVariantAlternative<T> = 1>
+            const T * get_if() const
+            {
+                if (holds_alternative<T>())
+                    return reinterpret_cast<T *>(m_value);
+                else
+                    return nullptr;
+            }
+
+            template<std::size_t Index>
+            const VariantAlternative<Index>& get() const
+            {
+                assert(Index == m_index);
+                return VariantAlternative<Index>(*this);
+            }
+
+            /* This is just a templated way to say
+             * "int*" for
+             * a VariantAlternative<0, Variant<int, char, long>()>*/
+            template< std::size_t Index >
+            using RawAlternativePointerT = typename std::add_pointer<VariantAlternative<Index>>::type;
+
+            template<std::size_t Index>
+            const RawAlternativePointerT<Index> * get_if() const
+            {
+                if (holds_alternative<Index>())
+                    return reinterpret_cast<RawAlternativePointerT<Index> *>(m_value);
+                else
+                    return nullptr;
+            }
+
+            std::size_t index() const
+            {
+                return m_index;
+            }
+
+            ~Variant()
+            {
+                Destroy();
+            }
+
+          private:
+            static constexpr std::size_t STORAGE_SIZE = VariantDetail::ParameterPackSize::GetMaxSizeOf<Ts...>();
+
+            alignas(Ts...) char m_storage[STORAGE_SIZE];
+            void *m_value = nullptr;
```

Comment:
debatable: is it worth having `m_value`? In Optional, we needed SOME extra variable to track whether or not it holds a value, but Variant is required to be holding a value. We could have used a bool. My memory's not good enough to know whether we chose a pointer over a bool for alignment reasons, or just because it made our code/debugging simpler.

But you could get away without using it at all for Variant, and just using reinterpret_casts

Reply:
Done, removed m_value, 
still in places where the result of reinterpret cast is returned - a temp local variable is used to satisfy old gcc versions complaining about type strict aliasing.